src/mame/barcrest/mpu1.cpp

```diff
@@ -35,85 +53,233 @@
 namespace {
 
 #include "mpu1.lh"
+#include "mpu2.lh"
 #include "m_bappl2.lh"
 #include "m_bapple.lh"
 #include "m_gndgit.lh"
 #include "m_lndg.lh"
 #include "m_mtchit.lh"
 #include "m_mtchup.lh"
-
-class mpu1_state : public driver_device
+#include "m2comet.lh"
+#include "m2frpoly.lh"
+#include "m2hilite.lh"
+#include "m2luckyl.lh"
+#include "m2luckys.lh"
+#include "m2rockon.lh"
+#include "m2splite.lh"
+#include "m2sstar.lh"
+#include "m2starl.lh"
+#include "m2svlite.lh"
+#include "m2triple.lh"
+
+class mpu12_base_state : public driver_device
 {
 public:
-	mpu1_state(const machine_config &mconfig, device_type type, const char *tag) :
+	mpu12_base_state(const machine_config &mconfig, device_type type, const char *tag) :
 		driver_device(mconfig, type, tag),
 		m_maincpu(*this, "maincpu"),
 		m_nmi_timer(*this, "nmi"),
 		m_pia1(*this, "pia1"),
 		m_pia2(*this, "pia2"),
 		m_lamps(*this, "lamp%u", 0U),
-		m_reels(*this, "emreel%u", 1U),
 		m_dac(*this, "dac"),
-		m_samples(*this, "samples")
+		m_samples(*this, "samples"),
+		m_pia3(*this, "pia3"),
+		m_pia4(*this, "pia4"),
+		m_nvram(*this, "nvram", 0x100, ENDIANNESS_BIG)
 	{ }
 
-	DECLARE_INPUT_CHANGED_MEMBER(coin_input);
+	void mpu12_base(machine_config &config);
```

Comment:
If this is an abstract class, this constructor and member function should be in the `protected` section.

Reply:
Fixed.